File: clients/go/internal/gate/record.go
package gate

import (
	"strings"

	"daisugi-verify/internal/pmodel"
	"daisugi-verify/internal/pyjson"
	"daisugi-verify/internal/pystr"
)

// applyPatchTool is OpenCode's apply_patch under the plugin's MCP-style
// name. It writes files, so it is never treated as an MCP call.
const applyPatchTool = "mcp__opencode__apply_patch"

var toolTypeMap = map[string]string{
	"Bash": "shell", "shell": "shell", "command": "shell",
	"Edit": "file_write", "Write": "file_write", "MultiEdit": "file_write",
	"Read": "file_read", "Glob": "file_read", "Grep": "file_read", "search": "file_read",
	"WebFetch": "network", "WebSearch": "network",
}

var piToolTypeMap = map[string]string{
	"bash": "shell", "powershell": "shell",
	"read": "file_read", "write": "file_write", "edit": "file_write",
}

// joinKeys is hook.JOIN_KEYS, in order.
var joinKeys = []string{
	"tool_use_id", "agent_id", "agent_type", "cwd", "transcript_path",
	"hook_event_name", "permission_mode",
}

// classifyTool is hook._classify_tool.
func classifyTool(name, fmt string) string {
	if strings.HasPrefix(name, "mcp__") {
		return "mcp"
	}
	if t, ok := toolTypeMap[name]; ok {
		return t
	}
	if fmt == "pi" {
		if t, ok := piToolTypeMap[name]; ok {
			return t
		}
		return "mcp"
	}
	return ""
}

// parseMCPToolName is hook._parse_mcp_tool_name.
func parseMCPToolName(name string) (string, string, bool) {
	if !strings.HasPrefix(name, "mcp__") {
		return "", "", false
	}
	server, tool, ok := strings.Cut(name[len("mcp__"):], "__")
	if !ok || server == "" || tool == "" {
		return "", "", false
	}
	return server, tool, true
}

// record is one normalized capture record, hook._payload_to_record's dict.
type record struct {
	ToolName string
	StepType string
	// SearchPaths are the paths a file read's input names under
	// file_path, path and filePath, in that order (search_rule.search_paths).
	// BadPath is true when one of them is truthy but not a string.
	SearchPaths []string
	BadPath     bool
	// Command, Path and URL are str(value or "") of the record's fields,
	// the form the rules and the log lines read. The *Raw fields are the
	// values themselves, which a step model validates.
	Command    string
	Path       string
	URL        string
	CommandRaw any
	PathRaw    any
	URLRaw     any
	MCPServer  string
	MCPTool    string
	Arguments  *pyjson.Object
	// obj is the record as Python builds it, key for key, with
	// captured_at set when it is written.
	obj *pyjson.Object
}

// toolNameOf is the payload's tool name: tool_name, else tool, else name,
// by Python truthiness.
func toolNameOf(p *pyjson.Object) any {
	for _, k := range []string{"tool_name", "tool", "name"} {
		if v := p.Value(k); pyjson.Truthy(v) {
			return v
		}
	}
	return nil
}

// toolInputOf is `payload.get("tool_input") or ... or {}`.
func toolInputOf(p *pyjson.Object) any {
	for _, k := range []string{"tool_input", "args", "input"} {
		if v := p.Value(k); pyjson.Truthy(v) {
			return v
		}
	}
	return pyjson.NewObject()
}

// firstRaw is `inp.get(a) or inp.get(b) or ... or ""`.
func firstRaw(inp *pyjson.Object, keys ...string) any {
	for _, k := range keys {
		if v := inp.Value(k); pyjson.Truthy(v) {
			return v
		}
	}
	return ""
}

// joinOf is hook.join_keys: the JOIN_KEYS the payload holds with a truthy
// value, in order.
func joinOf(p *pyjson.Object) *pyjson.Object {
	out := pyjson.NewObject()
	for _, k := range joinKeys {
		if v := p.Value(k); pyjson.Truthy(v) {
			out.Set(k, v)
		}
	}
	return out
}

// payloadToRecord is hook._payload_to_record. ok is false where Python
// returns None. It raises what Python raises on a field of an unexpected
// type (a tool name that is no str, a tool input that is no dict, content
// with no length).
func (r *runner) payloadToRecord(p *pyjson.Object, fmt string) (*record, bool) {
	tn := toolNameOf(p)
	if tn == nil {
		return nil, false
	}
	name, isStr := tn.(string)
	if !isStr {
		// _classify_tool: name.startswith("mcp__")
		noAttr(tn, "startswith")
	}
	step := classifyTool(name, fmt)
	if step == "" {
		return nil, false
	}
	inpAny := toolInputOf(p)
	rec := &record{ToolName: name, StepType: step, obj: pyjson.NewObject()}
	rec.obj.Set("captured_at", nil)
	rec.obj.Set("session_id", r.safeSession(p.Value("session_id")))
	rec.obj.Set("tool_name", name)
	rec.obj.Set("step_type", step)
	if step != "mcp" {
		inp, isObj := inpAny.(*pyjson.Object)
		if !isObj {
			noAttr(inpAny, "get")
		}
		switch step {
		case "shell":
			rec.CommandRaw = firstRaw(inp, "command", "cmd")
			rec.Command = pyStrOr(rec.CommandRaw)
			rec.obj.Set("command", rec.CommandRaw)
		case "file_read", "file_write":
			for _, k := range []string{"file_path", "path", "filePath"} {
				v := inp.Value(k)
				if !pyjson.Truthy(v) {
					continue
				}
				if sv, ok := v.(string); ok {
					rec.SearchPaths = append(rec.SearchPaths, sv)
				} else {
					rec.BadPath = true
				}
			}
			rec.PathRaw = firstRaw(inp, "file_path", "path", "filePath", "pattern")
			cwd, cwdIsStr := p.Value("cwd").(string)
			if ps, ok := rec.PathRaw.(string); ok && ps != "" && !isabs(ps) && !strings.HasPrefix(ps, "~") &&
				!strings.Contains(ps, "$") && cwdIsStr && isabs(cwd) {
				rec.PathRaw = normpath(join(cwd, ps))
			}
			rec.Path = pyStrOr(rec.PathRaw)
			rec.obj.Set("path", rec.PathRaw)
			if step == "file_write" {
				rec.obj.Set("content_len", pyLenOf(firstRaw(inp, "content", "new_string", "newString")))
			}
		case "network":
			rec.URLRaw = firstRaw(inp, "url", "query")
			rec.URL = pyStrOr(rec.URLRaw)
			rec.obj.Set("url", rec.URLRaw)
		}
	} else {
		if name == applyPatchTool {
			return nil, false
		}
		server, tool, ok := parseMCPToolName(name)
		if !ok {
			if fmt != "pi" {
				return nil, false
			}
			server, tool = "pi", name
		}
		rec.MCPServer, rec.MCPTool = server, tool
		args, isObj := inpAny.(*pyjson.Object)
		if !isObj {
			args = pyjson.NewObject()
		}
		rec.Arguments = args
		rec.obj.Set("mcp_server", server)
		rec.obj.Set("mcp_tool", tool)
		rec.obj.Set("arguments", args)
	}
	j := joinOf(p)
	for _, k := range j.Keys() {
		rec.obj.Set(k, j.Value(k))
	}
	return rec, true
}

// stepFieldCheck is _records_to_steps's step model on the record: the str
// field it validates, raising pydantic's error when the value is no str.
func (rec *record) stepFieldCheck() {
	var model, field string
	var v any
	switch rec.StepType {
	case "shell":
		model, field, v = "ShellStep", "command", rec.CommandRaw
	case "file_read":
		model, field, v = "FileReadStep", "path", rec.PathRaw
	case "file_write":
		model, field, v = "FileWriteStep", "path", rec.PathRaw
	case "network":
		model, field, v = "NetworkStep", "url", rec.URLRaw
	case "mcp":
		// ActionPlan's after-validator (models.non_finite_error): NaN or
		// an infinity in the call's arguments makes the plan invalid.
		if errs := pmodel.NonFinite(rec.Arguments, []any{"steps", 0, "arguments"}); len(errs) > 0 {
			e := &pmodel.ValidationError{Title: "ActionPlan", Errs: errs}
			panic(pystr.NewException("ValidationError", e.String()))
		}
		return
	default:
		return
	}
	if !pyjson.Truthy(v) {
		v = ""
	}
	if _, ok := v.(string); !ok {
		panic(stringTypeError(model, field, v))
	}
}

// pyStrings is the _strings helper of pane_rule and floor_config: every
// string inside a value, to depth 8.
func pyStrings(v any, depth int) []string {
	if depth > 8 {
		return nil
	}
	switch x := v.(type) {
	case string:
		return []string{x}
	case *pyjson.Object:
		var out []string
		for _, k := range x.Keys() {
			out = append(out, pyStrings(x.Value(k), depth+1)...)
		}
		return out
	case []any:
		var out []string
		for _, e := range x {
			out = append(out, pyStrings(e, depth+1)...)
		}
		return out
	}
	return nil
}

File: clients/rust/src/gate/record.rs
//! `hook._payload_to_record`: one hook payload as a normalized capture
//! record, with the payload's own values kept as they came, whatever their
//! JSON type, so every later step reads them as Python would.

use super::frames::frame;
use super::paths::{isabs, join2, normpath};
use super::py::PyErr;
use super::pyjson::{py_str_or_empty, py_type_name, Object, Value};
use super::pystr::{py_len, safe_session_id};
use super::{Runner, R};

/// OpenCode's apply_patch under the plugin's MCP-style name. It writes
/// files, so it is never treated as an MCP call.
pub const APPLY_PATCH_TOOL: &str = "mcp__opencode__apply_patch";

/// `hook.JOIN_KEYS`, in order.
pub const JOIN_KEYS: &[&str] =
    &["tool_use_id", "agent_id", "agent_type", "cwd", "transcript_path", "hook_event_name", "permission_mode"];

/// `hook._classify_tool`; `None` where Python returns None.
pub fn classify_tool(name: &str, fmt: &str) -> Option<&'static str> {
    if name.starts_with("mcp__") {
        return Some("mcp");
    }
    let t = match name {
        "Bash" | "shell" | "command" => Some("shell"),
        "Edit" | "Write" | "MultiEdit" => Some("file_write"),
        "Read" | "Glob" | "Grep" | "search" => Some("file_read"),
        "WebFetch" | "WebSearch" => Some("network"),
        _ => None,
    };
    if t.is_some() {
        return t;
    }
    if fmt == "pi" {
        return Some(match name {
            "bash" | "powershell" => "shell",
            "read" => "file_read",
            "write" | "edit" => "file_write",
            _ => "mcp",
        });
    }
    None
}

/// `hook._parse_mcp_tool_name`.
pub fn parse_mcp_tool_name(name: &str) -> Option<(String, String)> {
    let rest = name.strip_prefix("mcp__")?;
    let (server, tool) = rest.split_once("__")?;
    if server.is_empty() || tool.is_empty() {
        return None;
    }
    Some((server.to_string(), tool.to_string()))
}

/// One normalized capture record.
#[derive(Debug, Clone, Default)]
pub struct Record {
    /// The tool name (a str, or `_classify_tool` would have raised).
    pub tool_name: String,
    pub step_type: String,
    /// `str(record.get("command") or "")`, and the same for path and url:
    /// what the rules and the tier read.
    pub command: String,
    pub path: String,
    pub url: String,
    pub mcp_server: String,
    pub mcp_tool: String,
    pub arguments: Object,
    /// The record as Python builds it, key for key, with the payload's own
    /// values; captured_at is set when it is written.
    pub obj: Object,
}

impl Record {
    /// `record.get(k)`.
    pub fn raw(&self, k: &str) -> &Value {
        self.obj.value(k)
    }

    /// `str(record.get("command") or record.get("path") or "")`, the
    /// detail a hard-deny refusal names.
    pub fn rule_detail(&self) -> String {
        if !self.command.is_empty() {
            self.command.clone()
        } else {
            self.path.clone()
        }
    }

    /// `str(record.get("command") or record.get("path") or record.get("url")
    /// or "")`, the detail a verdict names.
    pub fn detail(&self) -> String {
        if !self.command.is_empty() {
            self.command.clone()
        } else if !self.path.is_empty() {
            self.path.clone()
        } else {
            self.url.clone()
        }
    }
}

/// The payload's tool name: tool_name, else tool, else name, by truthiness.
pub fn tool_name_of(p: &Object) -> Option<&Value> {
    ["tool_name", "tool", "name"].iter().map(|k| p.value(k)).find(|v| v.truthy())
}

/// `payload.get("tool_input") or payload.get("args") or ... or {}`.
pub fn tool_input_of(p: &Object) -> Value {
    for k in ["tool_input", "args", "input"] {
        let v = p.value(k);
        if v.truthy() {
            return v.clone();
        }
    }
    Value::Obj(Object::new())
}

/// `hook.join_keys`: the JOIN_KEYS the payload holds with a truthy value.
pub fn join_of(p: &Object) -> Object {
    let mut out = Object::new();
    for k in JOIN_KEYS {
        let v = p.value(k);
        if v.truthy() {
            out.set(k, v.clone());
        }
    }
    out
}

/// `hook._safe_session_id(raw)`: `str(raw or "")`, sanitized.
pub fn safe_session_value(v: &Value) -> R<String> {
    Ok(safe_session_id(&py_str_or_empty(v)?))
}

/// `inp.get(a) or inp.get(b) or ... or ""` on a dict.
fn first_truthy(inp: &Object, keys: &[&str]) -> Value {
    for k in keys {
        let v = inp.value(k);
        if v.truthy() {
            return v.clone();
        }
    }
    Value::Str(String::new())
}

/// `AttributeError` for `value.attr`, as Python words it.
fn no_attribute(v: &Value, attr: &str) -> PyErr {
    PyErr::attribute(format!("'{}' object has no attribute '{attr}'", py_type_name(v)))
}

/// `len(v)`.
pub fn py_len_of(v: &Value) -> Result<i64, PyErr> {
    match v {
        Value::Str(s) => Ok(py_len(s) as i64),
        Value::List(l) => Ok(l.len() as i64),
        Value::Obj(o) => Ok(o.len() as i64),
        other => Err(PyErr::type_err(format!("object of type '{}' has no len()", py_type_name(other)))),
    }
}

/// The dict a record step reads its fields from, or the AttributeError
/// Python raises on `inp.get`.
fn as_dict<'a>(inp: &'a Value) -> Result<&'a Object, PyErr> {
    match inp {
        Value::Obj(o) => Ok(o),
        other => Err(no_attribute(other, "get")),
    }
}

impl Runner {
    /// `hook._payload_to_record`. `None` where Python returns None.
    pub fn payload_to_record(&self, p: &Object, fmt: &str) -> R<Option<Record>> {
        let _f = frame();
        let name_v = match tool_name_of(p) {
            None => return Ok(None),
            Some(v) => v,
        };
        // _classify_tool calls name.startswith, which only a str has.
        let name = match name_v {
            Value::Str(s) => s.clone(),
            other => return Err(no_attribute(other, "startswith").into()),
        };
        let step = match classify_tool(&name, fmt) {
            None => return Ok(None),
            Some(s) => s,
        };
        let inp = tool_input_of(p);
        let mut rec = Record { tool_name: name.clone(), step_type: step.to_string(), ..Default::default() };
        rec.obj.set("captured_at", Value::Null);
        rec.obj.set("session_id", safe_session_value(p.value("session_id"))?);
        rec.obj.set("tool_name", name.as_str());
        rec.obj.set("step_type", step);
        match step {
            "shell" => {
                let d = as_dict(&inp)?;
                let v = first_truthy(d, &["command", "cmd"]);
                rec.command = py_str_or_empty(&v)?;
                rec.obj.set("command", v);
            }
            "file_read" | "file_write" => {
                let d = as_dict(&inp)?;
                let mut v = first_truthy(d, &["file_path", "path", "filePath", "pattern"]);
                if let (Value::Str(path), Value::Str(cwd)) = (&v, p.value("cwd")) {
                    if !path.is_empty()
                        && !isabs(path)
                        && !path.starts_with('~')
                        && !path.contains('$')
                        && isabs(cwd)
                    {
                        v = Value::Str(normpath(&join2(cwd, path)));
                    }
                }
                rec.path = py_str_or_empty(&v)?;
                rec.obj.set("path", v);
                if step == "file_write" {
                    let content = first_truthy(d, &["content", "new_string", "newString"]);
                    rec.obj.set("content_len", py_len_of(&content)?);
                }
            }
            "network" => {
                let d = as_dict(&inp)?;
                let v = first_truthy(d, &["url", "query"]);
                rec.url = py_str_or_empty(&v)?;
                rec.obj.set("url", v);
            }
            _ => {
                if name == APPLY_PATCH_TOOL {
                    return Ok(None);
                }
                let (server, tool) = match parse_mcp_tool_name(&name) {
                    Some(st) => st,
                    None if fmt == "pi" => ("pi".to_string(), name.clone()),
                    None => return Ok(None),
                };
                let args = match inp {
                    Value::Obj(o) => o,
                    _ => Object::new(),
                };
                rec.obj.set("mcp_server", server.as_str());
                rec.obj.set("mcp_tool", tool.as_str());
                rec.obj.set("arguments", args.clone());
                rec.mcp_server = server;
                rec.mcp_tool = tool;
                rec.arguments = args;
            }
        }
        let j = join_of(p);
        for k in j.keys() {
            rec.obj.set(k, j.value(k).clone());
        }
        Ok(Some(rec))
    }
}

/// The `_strings` helper of pane_rule and floor_config: every string inside
/// a value, to depth 8.
pub fn py_strings(v: &Value, depth: usize, out: &mut Vec<String>) {
    if depth > 8 {
        return;
    }
    match v {
        Value::Str(s) => out.push(s.clone()),
        Value::Obj(o) => {
            for k in o.keys() {
                py_strings(o.value(k), depth + 1, out);
            }
        }
        Value::List(l) => {
            for e in l {
                py_strings(e, depth + 1, out);
            }
        }
        _ => {}
    }
}
